### PR TITLE
Add jsonpath expand library

### PR DESCRIPTION
**Pull request by skaegi**
# Changes
This adds a "jsonpath" package with an "Expand" function. Functions in expand.go have some basic doc however for now the best examples of usage are in expand_test.go.

This is "expand" engine support that is part of https://github.com/tektoncd/pipeline/issues/1463

This is not currently user-facing and just a first step.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

### CODE REVIEW

File: pkg/jsonpath/expand_test.go [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+		{"$(string)", context["string"], false},
+		{"$(array)", context["array"], false},
+		{"$(obj)", context["obj"], false},
+		{"$(obj.a)", context["obj"].(map[string]interface{})["a"], false},
+		{"$(['obj'].a)", context["obj"].(map[string]interface{})["a"], false},
+		{"$(['obj']['a'])", context["obj"].(map[string]interface{})["a"], false},
+		{"$(obj['a'])", context["obj"].(map[string]interface{})["a"], false},
+		{"$(null)", context["null"], false},
+		{"$(array[*])", context["array"], true},
+		{"$(array[1:3])", []interface{}{context["array"].([]interface{})[1], context["array"].([]interface{})[2]}, true},
```
Comment (sbwsg) [inline pkg/jsonpath/expand_test.go]: Minor nit: suggest a helper func like:

```go
func arrLookup(context map[string]interface{}, i int) interface{} {
  return context["array"].([]interface{})[i]
}
```

Just to make this expected values here and line 151 a tad more readable.

File: pkg/jsonpath/expand.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	// This regex currently will only accept one-level of filter-expression e.g. nested (...) expressions but not (...(...))
+	expandRE = regexp.MustCompile(`\$\$|\$\((?:[^()]|\([^()]*\))+\)`)
+	// the list of valid Tekton JSONPath prefixes
+	expressionPrefix = regexp.MustCompile(`^[$.[@'"]`)
+)
+
+// createExpression takes a Tekton expression of the form $(...) and turns it into a hopefully valid Kubernets JSONPath expression
+func createExpression(variable string) string {
+	expression := strings.TrimSuffix(strings.TrimPrefix(variable, "$("), ")")
+	// we will dollar dot-prefix all expressions that don't already have a valid prefix
```
[the remark this answers is not in the document]
Comment (skaegi) [reply pkg/jsonpath/expand.go]: Thanks Scott, yes that is the intent and what I should document. Will add it.

File: pkg/jsonpath/expand.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	"regexp"
+	"strings"
+)
+
+var (
+	// expandRE captures the strings $$ (e.g. escaped dollar-sign) and those enclosed in $() (e.g. Tekton expression)
+	// This regex currently will only accept one-level of filter-expression e.g. nested (...) expressions but not (...(...))
+	expandRE = regexp.MustCompile(`\$\$|\$\((?:[^()]|\([^()]*\))+\)`)
+	// the list of valid Tekton JSONPath prefixes
+	expressionPrefix = regexp.MustCompile(`^[$.[@'"]`)
```
[the remark this answers is not in the document]
Comment (skaegi) [reply pkg/jsonpath/expand.go]: Correct... These are requirements from the Kube JSONPath library. I can doc this a bit better.

File: pkg/jsonpath/expand_test.go (hunk 1 above)
Comment (skaegi) [reply pkg/jsonpath/expand_test.go]: sure. will do.

File: pkg/jsonpath/expand.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	"encoding/json"
+	"fmt"
+	"k8s.io/client-go/util/jsonpath"
+	"regexp"
+	"strings"
+)
+
+var (
+	// expandRE captures the strings $$ (e.g. escaped dollar-sign) and those enclosed in $() (e.g. Tekton expression)
+	// This regex currently will only accept one-level of filter-expression e.g. nested (...) expressions but not (...(...))
```
[the remark this answers is not in the document]
Comment (skaegi) [reply pkg/jsonpath/expand.go]: A single level is for jsonpath filter expressions. Currently no need for a deeper-level.

**Pull request merged**